docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+
+PROTOTYPE: Write more here
+
+### Serialization
+
+**User Scenario**
+
+Serialization is often implemented using _dynamic analysis_, i.e. serializers often
+use reflection to examine the runtime state of a given type and generate serialization
+logic. This can be expensive and brittle. If the compile-time type and the runtime-type
+are similar, it could be useful to move much of the cost to compile-time, instead of
+run-time.
+
+Source generators provide a way to do this. Since source generators can be delivered via
+NuGet the same way analyzers can, we anticipate this would be a use-case for a source
+generator library, as opposed to everyone building their own.
+
+**Solution**
+
+To start, the generator will need some way to discover which types are meant
+to be serializable. One indicator could be an attribute, e.g.
+
+```C#
+[GeneratorSerializable]
+partial class MyRecord
+{
+    public string Item1 { get; }
+    public int Item2 { get; }
+}
+```
+
+This attribute could also be used for [Participate in the IDE experience](#participate-in-the-ide-experience),
+when the full scope of that feature is fully designed. In that scenario,
+instead of the generator finding every type marked with the given attribute,
+the compiler would notify the generator of every type marked with the given
+attribute. For now we'll assume that the types are provided to us.
+
+The first task is to decide what we want our serialization to return. Let's say
+we do a simple JSON serialization that produces a string like the following
+
+```json
+{
+    "Item1": "abc",
+    "Item2": 11,
+}
+```
+
+For that we could add a `Serialize` method to our record type like the following:
+
+```C#
+public string Serialize()
+{
+    var sb = new StringBuilder();
+    sb.AppendLine("{");
+    int indent = 8;
+
+    // Body
+    addWithIndent($"\"Item1\": \"{this.Item1.ToString()}\",");
+    addWithIndent($"\"Item2\": {this.Item2.ToString()},");
+
+    sb.AppendLine("}");
+
+    return sb.ToString();
+
+    void addWithIndent(string s)
+    {
+        sb.Append(' ', indent);
+        sb.AppendLine(s);
+    }
+}
+```
+
+Obviously this is heavily simplified -- this example only handles the `string` and `int`
+types properly, adds a trailing comma to the json output and has no error recovery, but
+it should serve to demonstrate the kind of code a source generator could add to a compilation.
+
+Our next task is design a generator to generate the above code, since the
+above code is itself customized in the `// Body` section according to the
+actual properties in the class. In other words, we need to generate the code
+which will generate the JSON format. This is a generator-generator.
+
+Let's start with a basic template. We are adding a full source generator, so we'll need
+to generate a class with the same name as the input class, with a public method called
+`Serialize`, and a filler area where we write out the properties.
+
+```C#
+string template = @"
```

Comment:
Shameless plug: do we want to officially suggesting using [PolySharp](https://github.com/Sergio0694/PolySharp) in source generators, so folks can still enable C# 12 and at least remove a whole lot of friction writing them? Eg. here we could show using an interpolated raw multiline string literal as well, and we could use C# 12 syntax whenever applicable in other places too?